### ISSUE TITLE
Scripting: layer instanceof TileLayer produces a type error

### ISSUE THREAD

**Issue by yanchith** [mapeditor/tiled#3232]
Hey :)

Thanks for making Tiled ❤️ 

When implementing an exporter to a custom level format from Tiled's `TileMap`, I managed to produce a type error (`TypeError: Type error`) from the script when doing:

```js
const isTileLayer = layer instanceof TileLayer; // For the record, layer really was a TileLayer
```

In the `d.ts`, `TileLayer` is defined as:

```js
declare class TileLayer extends Layer
```

.. but I imagine this is probably not a JS class, but Tiled's data structure that is just accessible from JS, but it likely doesn't implement all the operators.

I know, there's `Layer.isTileLayer`, but this probably shouldn't crash.

Platform: Windows
Tiled Version: 1.7.2

**Comment by bjorn** [comment]
> .. but I imagine this is probably not a JS class, but Tiled's data structure that is just accessible from JS, but it likely doesn't implement all the operators.

You're right, the `layer` is not actually an instance of `TileLayer`. It is a JS object wrapping an instance of the C++ class `Tiled::EditableTileLayer`. And the `TileLayer` variable is not actually a class either. It is a JS object wrapping the "metaobject" of that C++ class, which allows calling the constructor and accessing enums (though this class doesn't have any, but for example like `TileMap.Orthogonal`).

So, this whole setup has very little to do with how things normally work in JavaScript, and I think `instanceof` raising a type error is probably the correct thing to do in this case. I've marked this as a Qt issue, but I'm not sure if it can be fixed. Are you saying it may be possible to make this work by implementing some operators?

Regarding the TypeScript definitions, would it help if they were defined using interfaces instead of class definitions, or do interfaces raise the same expectations regarding `instanceof`?

**Comment by bjorn** [comment]
I found [QTBUG-79868](https://bugreports.qt.io/browse/QTBUG-79868) regarding the throwing of `TypeError`, which was apparently fixed to return `false` instead. Could be interesting to look into why we're still getting a `TypeError` in this case. [This function](https://code.woboq.org/qt6/qtdeclarative/src/qml/qml/qqmltypewrapper.cpp.html#_ZN3QV415QQmlTypeWrapper17virtualInstanceOfEPKNS_6ObjectERKNS_5ValueE) might be related.

**Comment by yanchith** [comment]
> So, this whole setup has very little to do with how things normally work in JavaScript, and I think instanceof raising a type error is probably the correct thing to do in this case. I've marked this as a Qt issue, but I'm not sure if it can be fixed. Are you saying it may be possible to make this work by implementing some operators?

Sorry, I don't actually have real experience with integrating with QT's javascript VM, I just assumed there might be something like that, and [that function](https://code.woboq.org/qt6/qtdeclarative/src/qml/qml/qqmltypewrapper.cpp.html#_ZN3QV415QQmlTypeWrapper17virtualInstanceOfEPKNS_6ObjectERKNS_5ValueE) you mentioned seems to be the place to look.

> Regarding the TypeScript definitions, would it help if they were defined using interfaces instead of class definitions, or do interfaces raise the same expectations regarding instanceof?

Since interfaces do not actually exist in JS, there should be no expectations of `instanceof` reporting anything useful. It still probably shouldn't crash, but less people would try :)
